File: java/spring-hibernate/src/main/java/com/sprint/sms/api/access/RoleAccess.java
package com.sprint.sms.api.access;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

import com.ideahut.common2.annotation.IdhFormatter;

@IdhFormatter
public class RoleAccess implements Serializable {
	
	/**
	 * 
	 */
	private static final long serialVersionUID = 2627232198404269622L;
	
	private Long id;
	
	private String name;

	private List<MenuAccess> menuList = new ArrayList<MenuAccess>();
	
	private Map<Long, MenuAccess> menuMap = new HashMap<Long, MenuAccess>();
	
	private Map<String, Set<String>> pathMap = new HashMap<String, Set<String>>(); // dipisahkan dengan parent path agar pencarian lebih baik :))

	@IdhFormatter
	public Long getId() {
		return id;
	}

	public void setId(Long id) {
		this.id = id;
	}

	@IdhFormatter
	public String getName() {
		return name;
	}

	public void setName(String name) {
		this.name = name;
	}

	@IdhFormatter
	public List<MenuAccess> getMenuList() {
		return menuList;
	}

	public void setMenuList(List<MenuAccess> menuList) {
		this.menuList = menuList;
	}

	@IdhFormatter
	public Map<Long, MenuAccess> getMenuMap() {
		return menuMap;
	}

	public void setMenuMap(Map<Long, MenuAccess> menuMap) {
		this.menuMap = menuMap;
	}

	@IdhFormatter
	public Map<String, Set<String>> getPathMap() {
		return pathMap;
	}

	public void setPathMap(Map<String, Set<String>> pathMap) {
		this.pathMap = pathMap;
	}

	public boolean isAllowedPath(String path) {
		if (path == null || path.length() == 0) {
			return false;
		}
		if (!path.startsWith("/")) {
			path = "/" + path;
		}
		String parent = path.substring(1);
		int pos = parent.indexOf("/");
		if (pos != -1) {
			parent = parent.substring(0, pos);
		}
		parent = "/" + parent;
		Set<String> pathSet = pathMap.get(parent);
		if (pathSet == null) {
			return false;
		}
		if (pathSet.isEmpty()) {
			return path.equals(parent);
		}
		return pathSet.contains(path);
	}
	
}

File: php/slim-doctrine/app/src/com/sprint/sms/api/access/RoleAccess.php
<?php
namespace App\com\sprint\sms\api\access;

use \App\com\sprint\sms\api\support\FormatSupport;

class RoleAccess implements FormatSupport
{
	private $id;
	
	private $name;

	private $menuList = array();
	
	private $menuMap = array();
	
	private $pathMap = array();
	
	
	public function getId() {
		return $this->id;
	}

	public function setId($id) {
		$this->id = $id;
	}

	public function getName() {
		return $this->name;
	}

	public function setName($name) {
		$this->name = $name;
	}

	public function getMenuList() {
		return $this->menuList;
	}

	public function setMenuList(array $menuList = array()) {
		$this->menuList = $menuList;
	}

	public function getMenuMap() {
		return $this->menuMap;
	}

	public function setMenuMap(array $menuMap = array()) {
		$this->menuMap = $menuMap;
	}

	public function getPathMap() {
		return $this->pathMap;
	}

	public function setPathMap(array $pathMap = array()) {
		$this->pathMap = $pathMap;
	}

	public function isAllowAccess($path) {
		if ($path === null || strlen($path) === 0) {
			return false;
		}
		$bos = substr($path, 0, 1) === "/";
		if (!$bos) {
			$path = "/" . $path;
		}
		$parent = substr($path, 1, strlen($path));
		$pos = strpos($parent, "/", 0);
		if ($pos) {
			$parent = substr($parent, 0, $pos);
		}
		$parent = "/" . $parent;
		if (!isset($this->pathMap[$parent])) {
			return false;
		}
		$pathSet = $this->pathMap[$parent];	
		if (count($pathSet) === 0) {
			return $path === $parent;
		}
		return array_search($path, $pathSet) !== false;
	}
	
	public function toFormatObject() {
		$o = array();
		$o["id"] = $this->id;
		$o["name"] = $this->name;
		$o["menuList"] = $this->menuList;
		$o["menuMap"] = $this->menuMap;	
		$o["pathMap"] = $this->pathMap;
		return $o;
	}
	
}
